networkx/algorithms/bipartite/EFM_algorithms.py

```diff
@@ -0,0 +1,291 @@
+
+        Like presented in the article, Y-path-saturated graph contains an empty envy-free matching so X_L and Y_L are empty in the partition.
+
+    """
+    if M is None:
+        M = nx.bipartite.maximum_matching(G, top_nodes=top_nodes)
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_subsets, Y_subsets = __M_alternating_sequence__(G, M, top_nodes)
+    X_S = set()
+    for subset in X_subsets:
+        X_S.update(subset)
+
+    Y_S = set()
+    for subset in Y_subsets:
+        Y_S.update(subset)
+
+    return set(X) - X_S, X_S, set(Y) - Y_S, Y_S
+
+
+@not_implemented_for("directed", "multigraph")
+def envy_free_matching(G, top_nodes=None):
+    r"""Return an envy-free matching of maximum cardinality
+    Parameters
+    ----------
+    G:
+        NetworkX graph
+        Undirected bipartite graph
+    top_nodes:
+                list
+                if graph is not all connected top_nodes is the set of the top nodes in the bipartite graph G
+    Returns
+    -------
+    Matching: dictionary
+        The Maximum cardinallity envy-free matching is returned as a dictionary.
+    References
+    ----------
+    .. [1] "Envy-free Matchings in Bipartite Graphs and their Applications to Fair Division",
+    by Elad Aigner-Horev and Erel Segal-Halevi (2022), https://arxiv.org/abs/1901.09527
+    Algorithm 2: Finding an envy-free matching of maximum cardinality.
+
+    Examples
+    --------
+        Example 1: Perfect matching
+        >>> Graph=nx.complete_bipartite_graph(3,3)
+        >>> envy_free_matching(Graph)
+        {0: 3, 1: 4, 2: 5, 3: 0, 4: 1, 5: 2}
+
+        Where there exists a perfect matching the maximum envy free matching is the perfect matching.
+        Example 2: Non-empty envy-free matching
+        >>> Graph=nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+        >>> envy_free_matching(Graph)
+        {0: 3, 3: 0}
+
+        Example 3: Odd path
+        >>> Graph=nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+        >>> envy_free_matching(Graph)
+        {}
+
+        Like presented in the article, odd path contains an empty envy-free matching so the returned matching is empty.
+        Example 4: Y-path-saturated graph
+        >>> Graph=nx.Graph([(0,6),(6,0),(1,6),(6,1),(1,7),(7,1),(2,6),(6,2),(2,8),(8,2),(3,9),(9,3),(3,6),(6,3),(4,8),(8,4),(4,7),(7,4),(5,9),(9,5)])
+        >>> envy_free_matching(Graph)
+        {}
+
+        Like presented in the article, Y-path-saturated graph contains an empty envy-free matching so X_L and Y_L are empty in the partition.
+    """
+    M = nx.bipartite.maximum_matching(G, top_nodes=top_nodes)
+    EFM_PARTITION = _EFM_partition(G, M, top_nodes)
+    un = EFM_PARTITION[1].union(EFM_PARTITION[3])
+    M = {node: M[node] for node in M if node not in un and M[node] not in un}
+    return M
+
+
+@not_implemented_for("directed", "multigraph")
+def minimum_weight_envy_free_matching(G, top_nodes=None):
```

Comment:
The name of this function should indicate that it returns an envy-free matching of maximum cardinality. Perhaps "min_weight_largest_envy_free_matching" or "min_weight_max_size_envy_free_matching".